helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -960,10 +956,10 @@ public void process(WatchedEvent event) {
 
   private void fireAllEvents() {
     for (Entry<String, Set<IZkChildListener>> entry : _childListener.entrySet()) {
-      fireChildChangedEvents(entry.getKey(), entry.getValue());
+      fireChildChangedEvents(entry.getKey(), entry.getValue(), true);
```

Comment:
So while the client is reconnected, it will just reinstall the watcher without checking, right? Will we still see watcher leakage?

Reply:
Good question. I think with the default value, it will reinstall the watcher whatever left on the previous session; It's likely the node deletion happens after session expiry, in this case, it will leak; One mitigation approach is to perform EXISTS check when handling new session, but it's always possible a race-condition exists (that the node deletion happen after the exists check).  One good thing though comes with new session is that all the "leaked" watchers will be gone.

Do you think a set of exists check on each path from original session should be performed or it's OK to trade-off between performance and excessive watchers.

Reply:
1. I still think the fix should be bumping up ZK version. If the concern is not super urgent, we should do that. Clean and no side effects.
2. Otherwise, given this issue, a background GC thread would be easier. What I'm thinking is that we have a running timeout for the existing check-in that thread for each watched path.

- If delete event, timeout set to zero and trigger GC immediately.
- Otherwise, the GC thread will validate the path after the timeout.

Reply:
1. I don't think to bump up the ZK version and use the new API to delete ZK watcher explicitly would resolve the problem. 

Let me explain why: if the node is re-created after deleting the watcher request is submitted, without the parent path's child listener, there's no way the ZkClient would be able to receive new events from the path. It's the same dilemma this fix is currently facing.

2. A background GC thread is an over-kill solution. First It's difficult to determine the right TTL; Second, it comes with the additional thread cost.

My argument point is that it's not always valid to say we need to be 100% stick to the old behavior, especially when the old behavior is incorrect and leading to the excessive amount of leakage.
The easiest way is to release a newer Helix version and educate the client with the caveat **"Once a node is deleted, the listener on the path will be inactive even the node is re-created; The workaround is to have a child listener on the parent path or re-subscribe the path"**. The statement IMO makes sense to most clients. 

As to how many clients will be suffering from the behavior change, I don't have a good answer but the experience told me the number is minor if the client is using Helix as a cluster management tool instead of a generic ZKClient proxy. 

Do these all make sense? @kaisun2000 @dasahcc please feel free to leave any thoughts about this, cause it will change the design fundamentally and won't be able to help Venice reduce the watcher pains in any near future.

Reply:
The first point is for us to simplify the GC. I agree it won't help with the listener becoming inactive case.
And, about the second part, a behavior change is fine as long as it is reasonable. However, the main concern is that, even with the current design, leakage is still possible, right? Since the watcher removing depends on the notification, if no notification arrived, or the client missed, the watcher will be leaked. That's why I think this design is not good enough. Feel free to fix the current design if there is a way to avoid all leakages.

Reply:
If I understand correctly, @jiajunwang concern is that reconnect happen and fireAllEvent would be triggered and install the watcher for all the paths in subscriber list. The thing is that there can be case that in the disconnecting period, the zkpath can be deleted. This can still leak another path. This is very good point. 

The first propose fix by @jiajunwang is:
1/ use exists to re-subscribe in re-connecting phase anyway.
2/ if path is not there, delete the watcher for this path.
This require 3.5 zk feature though. Not feasible now.

For the GC approach, what I can see is that this may still be a problem. First you check if the subscribe path is there. If it is not there, you delete it. The problem is that these are two steps operation to ZK. The problem is between the two steps, the user may create the path and install watcher as one operation. Now you delete it. This is even worse than leaking. 

The second step of the first proposal has the same problem. 

Actually I thought about this before, if Zookeeper does not have atomic check if the path not there then delete watcher, it is hard for us to implement this correctly.
@jiajunwang, what is your take?

Maybe let us note this issue and have an enhancement task to track it?

Reply:
@jiajunwang I agree that during re-connecting, it's still possible to leak the watcher if the in-memory listeners contain already deleted paths; I will think about the case and see if it can be overcome. But we all know at least the case here is not the major leakage point.

In fact, the long comment is more about the other one you claimed about changing the existing behavior cause I have to modify the integration test case.

Reply:
Discussed offline and we reached the consensus that the leakage during session expiry is a potential issue and since the impact of it is minor, we'll create a ticket to track the fixing progress in the future